src/sundials/sundatanode/sundatanode_inmem.c

```diff
@@ -0,0 +1,475 @@
+
+  *child_node = NULL;
+
+  sunbooleantype has_children;
+  SUNCheckCall(SUNDataNode_HasChildren_InMem(self, &has_children));
+
+  if (has_children)
+  {
+    if (SUNHashMap_GetValue(IMPL_MEMBER(self, named_children), name,
+                            (void**)child_node))
+    {
+      return SUN_ERR_DATANODE_NODENOTFOUND;
+    }
+  }
+
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNDataNode_RemoveChild_InMem(SUNDataNode self, sundataindex index,
+                                         SUNDataNode* child_node)
+{
+  SUNFunctionBegin(self->sunctx);
+
+  sunbooleantype has_children;
+  SUNCheckCall(SUNDataNode_HasChildren_InMem(self, &has_children));
+
+  if (!has_children) { return SUN_SUCCESS; }
+
+  SUNDataNode* child_node_ptr =
+    SUNStlVector_SUNDataNode_At(IMPL_MEMBER(self, anon_children), index);
+  if (child_node_ptr)
+  {
+    *child_node = *child_node_ptr;
+    if (*child_node)
+    {
+      IMPL_MEMBER(*child_node, parent) = NULL;
+      SUNStlVector_SUNDataNode_Erase(IMPL_MEMBER(self, anon_children), index);
+    }
+  }
+
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNDataNode_RemoveNamedChild_InMem(const SUNDataNode self,
+                                              const char* name,
+                                              SUNDataNode* child_node)
+{
+  SUNFunctionBegin(self->sunctx);
+
+  *child_node = NULL;
+
+  sunbooleantype has_children;
+  SUNCheckCall(SUNDataNode_HasChildren_InMem(self, &has_children));
+
+  if (has_children)
+  {
+    if (SUNHashMap_Remove(IMPL_MEMBER(self, named_children), name,
+                          (void**)child_node))
+    {
+      return SUN_ERR_DATANODE_NODENOTFOUND;
+    }
+    IMPL_MEMBER(*child_node, parent) = NULL;
+    IMPL_MEMBER(self, num_named_children)--;
+  }
+
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNDataNode_GetData_InMem(const SUNDataNode self, void** data,
+                                     size_t* data_stride, size_t* data_bytes)
+{
+  SUNFunctionBegin(self->sunctx);
+
+  SUNMemory leaf_data = (SUNMemory)IMPL_MEMBER(self, leaf_data);
+
+  *data_stride = leaf_data->stride;
+  *data_bytes  = leaf_data->bytes;
+  *data        = leaf_data->ptr;
+
+  return SUN_SUCCESS;
+}
+
+SUNErrCode SUNDataNode_GetDataNvector_InMem(const SUNDataNode self, N_Vector v,
+                                            sunrealtype* t)
+{
+  SUNFunctionBegin(self->sunctx);
+
+  void* queue = NULL;
+
+  SUNMemory leaf_data = (SUNMemory)IMPL_MEMBER(self, leaf_data);
+
+  SUNMemoryType leaf_mem_type = leaf_data->type;
+  SUNMemoryType buffer_mem_type = N_VGetDeviceArrayPointer(v) ? SUNMEMTYPE_DEVICE
+                                                              : SUNMEMTYPE_HOST;
+
+  sunindextype buffer_size = 0;
+  SUNCheckCall(N_VBufSize(v, &buffer_size));
+  SUNAssert((buffer_size + sizeof(sunrealtype)) == leaf_data->bytes,
+            SUN_ERR_ARG_INCOMPATIBLE);
+
+  if (leaf_mem_type == buffer_mem_type)
+  {
+    sunrealtype* data_ptr = leaf_data->ptr;
+    *t                    = data_ptr[0];
```

Comment:
Here and below, can we assume `data_ptr` is host data or should this use the memory helper copy?

Reply:
I'm missing it, where is this defined as host data?

Reply:
I see it now in `SUNDataNode_SetDataNvector_InMem`